File: demo-app-objc/CloudXObjCRemotePods/Views/Ad/BaseAdViewController.m
#import <UIKit/UIKit.h>
#import <dispatch/dispatch.h>
#import "BaseAdViewController.h"
#import <CloudXCore/CloudXCore.h>
#import <CloudXCore/CLXUserDefaultsKeys.h>
#import "LogsModalViewController.h"
#import "DemoAppLogger.h"
#import "CLXDemoConfigManager.h"
#import "DemoToastView.h"

@implementation BaseAdViewController

- (NSString *)appKey {
    return [[CLXDemoConfigManager sharedManager] currentConfig].appKey;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupStatusUI];
    [self setupShowLogsButton];
    [self updateStatusUIWithState:AdStateNoAd];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    // Clear logs when switching between different ad formats (tabs)
    NSString *currentAdFormat = NSStringFromClass([self class]);
    static NSString *lastAdFormat = nil;
    
    if (lastAdFormat && ![lastAdFormat isEqualToString:currentAdFormat]) {
        // Switching between different ad formats - clear logs for clean slate
        [[DemoAppLogger sharedInstance] clearLogs];
        [[DemoAppLogger sharedInstance] logMessage:[NSString stringWithFormat:@"[%@] Switched from %@ - logs cleared", currentAdFormat, lastAdFormat]];
    }
    // No log for same format - keep it clean
    
    // Remember current ad format for next time (session only)
    lastAdFormat = currentAdFormat;
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    // Don't clear logs when leaving - let user see the complete ad lifecycle
}

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message {
    NSString *safeTitle = title ?: @"Alert";
    NSString *safeMessage = message ?: @"";
    [DemoToastView showInViewController:self title:safeTitle message:safeMessage];
}

- (void)initializeSDKWithCompletion:(void (^)(BOOL success, NSError *error))completion {
    NSString *appKey = [self appKey];
    if (!appKey || [appKey length] == 0) {
        if (completion) completion(NO, [NSError errorWithDomain:@"CloudX" code:1 userInfo:@{NSLocalizedDescriptionKey: @"API key is missing."}]);
        return;
    }

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        CLXDemoConfig *config = [[CLXDemoConfigManager sharedManager] currentConfig];
        NSString *hashedUserId = config.hashedUserId;
        
        // Set hashed user ID before initialization if provided
        if (hashedUserId.length > 0) {
            [[CloudXCore shared] setHashedUserID:hashedUserId];
        }
        
        [self updateStatusUIWithState:AdStateLoading];
        // Use standard CloudXCore initialization with configuration object
        CLXInitializationConfiguration *initConfig = [CLXInitializationConfiguration configurationWithAppKey:appKey];
        [[CloudXCore shared] initializeWithConfiguration:initConfig completion:^(CLXSdkConfiguration * _Nullable sdkConfig, CLXError * _Nullable error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                BOOL success = (sdkConfig != nil);
                if (success) {
                    [[NSNotificationCenter defaultCenter] postNotificationName:@"cloudXSDKInitialized" object:nil];
                }
                if (completion) completion(success, error);
            });
        }];
    });
}

- (void)initializeSDK {
    [self initializeSDKWithCompletion:^(BOOL success, NSError * _Nullable error) {
        // No action needed for this convenience method
    }];
}

- (void)updateStatusUIWithState:(AdState)state {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *text;
        UIColor *color;
        
        switch (state) {
            case AdStateNoAd:
                text = @"No Ad Loaded";
                color = [UIColor systemRedColor];
                break;
            case AdStateLoading:
                text = @"Loading Ad...";
                color = [UIColor systemYellowColor];
                break;
            case AdStateReady:
                text = @"Ad Ready";
                color = [UIColor systemGreenColor];
                break;
        }
        
        self.statusLabel.text = text;
        self.statusLabel.textColor = color;
        self.statusIndicator.backgroundColor = color;
    });
}

- (void)setupStatusUI {
    // Setup status indicator stack
    self.statusStack = [[UIStackView alloc] init];
    self.statusStack.axis = UILayoutConstraintAxisHorizontal;
    self.statusStack.spacing = 8;
    self.statusStack.alignment = UIStackViewAlignmentCenter;
    self.statusStack.translatesAutoresizingMaskIntoConstraints = NO;
    
    self.statusIndicator = [[UIView alloc] init];
    self.statusIndicator.translatesAutoresizingMaskIntoConstraints = NO;
    self.statusIndicator.layer.cornerRadius = 6;
    self.statusIndicator.clipsToBounds = YES;
    
    self.statusLabel = [[UILabel alloc] init];
    self.statusLabel.textAlignment = NSTextAlignmentCenter;
    self.statusLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightMedium];
    self.statusLabel.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.statusStack addArrangedSubview:self.statusIndicator];
    [self.statusStack addArrangedSubview:self.statusLabel];
    
    [self.view addSubview:self.statusStack];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.statusStack.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.statusStack.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-20],
        [self.statusIndicator.widthAnchor constraintEqualToConstant:12],
        [self.statusIndicator.heightAnchor constraintEqualToConstant:12]
    ]];
}

- (void)setupCenteredButtonWithTitle:(NSString *)title action:(SEL)action {
    if (!title) title = @"Button";
    UIButton *button = [UIButton buttonWithType:UIButtonTypeSystem];
    [button setTitle:title forState:UIControlStateNormal];
    button.titleLabel.font = [UIFont boldSystemFontOfSize:16];
    button.backgroundColor = [UIColor systemBlueColor];
    [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    button.layer.cornerRadius = 8;
    button.translatesAutoresizingMaskIntoConstraints = NO;
    
    // Use UIButtonConfiguration for iOS 15+ to avoid deprecated contentEdgeInsets
    if (@available(iOS 15.0, *)) {
        UIButtonConfiguration *config = [UIButtonConfiguration filledButtonConfiguration];
        config.title = title;
        config.baseBackgroundColor = [UIColor systemBlueColor];
        config.baseForegroundColor = [UIColor whiteColor];
        config.cornerStyle = UIButtonConfigurationCornerStyleMedium;
        config.contentInsets = NSDirectionalEdgeInsetsMake(12, 24, 12, 24);
        button.configuration = config;
    } else {
        #pragma clang diagnostic push
        #pragma clang diagnostic ignored "-Wdeprecated-declarations"
        button.contentEdgeInsets = UIEdgeInsetsMake(12, 24, 12, 24);
        #pragma clang diagnostic pop
    }
    
    if (action) {
        [button addTarget:self action:action forControlEvents:UIControlEventTouchUpInside];
    }
    
    [self.view addSubview:button];
    
    [NSLayoutConstraint activateConstraints:@[
        [button.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [button.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor],
        [button.widthAnchor constraintEqualToConstant:200],
        [button.heightAnchor constraintEqualToConstant:44]
    ]];
}

- (void)setupShowLogsButton {
    // App Logs button (shows demo app logs)
    UIButton *showLogsButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [showLogsButton setTitle:@"App Logs" forState:UIControlStateNormal];
    showLogsButton.titleLabel.font = [UIFont systemFontOfSize:14 weight:UIFontWeightMedium];
    showLogsButton.backgroundColor = [UIColor systemOrangeColor];
    [showLogsButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    showLogsButton.layer.cornerRadius = 6;
    showLogsButton.translatesAutoresizingMaskIntoConstraints = NO;
    
    [showLogsButton addTarget:self action:@selector(showLogsModal) forControlEvents:UIControlEventTouchUpInside];
    
    [self.view addSubview:showLogsButton];
    
    // SDK Debugger button (toggles visual debugging)
    self.sdkDebuggerButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self updateSDKDebuggerButtonTitle];
    self.sdkDebuggerButton.titleLabel.font = [UIFont systemFontOfSize:14 weight:UIFontWeightMedium];
    self.sdkDebuggerButton.layer.cornerRadius = 6;
    self.sdkDebuggerButton.translatesAutoresizingMaskIntoConstraints = NO;
    
    [self.sdkDebuggerButton addTarget:self action:@selector(toggleSDKDebugger) forControlEvents:UIControlEventTouchUpInside];
    
    [self.view addSubview:self.sdkDebuggerButton];
    
    [NSLayoutConstraint activateConstraints:@[
        [showLogsButton.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:20],
        [showLogsButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20],
        [showLogsButton.widthAnchor constraintEqualToConstant:100],
        [showLogsButton.heightAnchor constraintEqualToConstant:32],
        
        [self.sdkDebuggerButton.topAnchor constraintEqualToAnchor:showLogsButton.bottomAnchor constant:8],
        [self.sdkDebuggerButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20],
        [self.sdkDebuggerButton.widthAnchor constraintEqualToConstant:100],
        [self.sdkDebuggerButton.heightAnchor constraintEqualToConstant:32]
    ]];
}

- (void)updateSDKDebuggerButtonTitle {
    BOOL isEnabled = [CloudXCore isVisualDebuggingEnabled];
    if (isEnabled) {
        [self.sdkDebuggerButton setTitle:@"Debugger ✓" forState:UIControlStateNormal];
        self.sdkDebuggerButton.backgroundColor = [UIColor systemGreenColor];
    } else {
        [self.sdkDebuggerButton setTitle:@"Debugger" forState:UIControlStateNormal];
        self.sdkDebuggerButton.backgroundColor = [UIColor systemPurpleColor];
    }
    [self.sdkDebuggerButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
}

- (void)toggleSDKDebugger {
    BOOL currentState = [CloudXCore isVisualDebuggingEnabled];
    [CloudXCore setVisualDebuggingEnabled:!currentState];
    [self updateSDKDebuggerButtonTitle];
}

- (void)showLogsModal {
    LogsModalViewController *logsModal = [[LogsModalViewController alloc] initWithTitle:@"Demo App Logs"];
    logsModal.modalPresentationStyle = UIModalPresentationPageSheet;
    
    [self presentViewController:logsModal animated:YES completion:nil];
}


@end 

File: demo-app-swift/CloudXSwiftRemotePods/Views/Ad/BaseAdViewController.swift
import UIKit
import CloudXCore

protocol AdStateManaging {
    var isLoading: Bool { get set }
    func updateStatusUI(state: AdState)
}

enum AdState {
    case noAd
    case loading
    case ready
    
    var text: String {
        switch self {
        case .noAd: return "No Ad Loaded"
        case .loading: return "Loading Ad..."
        case .ready: return "Ad Ready"
        }
    }
    
    var color: UIColor {
        switch self {
        case .noAd: return .systemRed
        case .loading: return .systemYellow
        case .ready: return .systemGreen
        }
    }
}

class BaseAdViewController: UIViewController, AdStateManaging {
    var cloudX: CloudXCore { CloudXCore.shared }
    var isLoading = false
    
    var appKey: String? {
        return CLXDemoConfigManager.sharedManager.currentConfig.appKey
    }
    
    let statusLabel: UILabel = {
        let label = UILabel()
        label.textAlignment = .center
        label.font = .systemFont(ofSize: 16, weight: .medium)
        label.translatesAutoresizingMaskIntoConstraints = false
        return label
    }()
    
    let statusIndicator: UIView = {
        let view = UIView()
        view.translatesAutoresizingMaskIntoConstraints = false
        view.layer.cornerRadius = 6
        view.clipsToBounds = true
        return view
    }()
    
    let statusStack: UIStackView = {
        let stack = UIStackView()
        stack.axis = .horizontal
        stack.spacing = 8
        stack.alignment = .center
        stack.translatesAutoresizingMaskIntoConstraints = false
        return stack
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupStatusUI()
        setupShowLogsButton()
        updateStatusUI(state: .noAd)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        // Clear logs when switching between different ad formats (tabs)
        let currentAdFormat = String(describing: type(of: self))
        
        if let lastFormat = BaseAdViewController.lastAdFormat, lastFormat != currentAdFormat {
            // Switching between different ad formats - clear logs for clean slate
            DemoAppLogger.sharedInstance.clearLogs()
            DemoAppLogger.sharedInstance.logMessage("[\(currentAdFormat)] Switched from \(lastFormat) - logs cleared")
        }
        // No log for same format - keep it clean
        
        // Remember current ad format for next time (session only)
        BaseAdViewController.lastAdFormat = currentAdFormat
    }
    
    // Static variable to track last ad format across all instances
    private static var lastAdFormat: String?
    
    func showAlert(title: String, message: String) {
        DispatchQueue.main.async {
            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
            alert.addAction(.init(title: "OK", style: .default))
            self.present(alert, animated: true)
        }
    }
    
    func initializeSDK() async {
        guard let appKey = appKey, !appKey.isEmpty else {
            showAlert(title: "Error", message: "API key is missing.")
            return
        }

        let config = CLXDemoConfigManager.sharedManager.currentConfig
        UserDefaults.standard.set(config.baseURL, forKey: "CloudXInitURL")
        
        // Set hashed user ID before initialization if provided
        if !config.hashedUserId.isEmpty {
            cloudX.setHashedUserID(config.hashedUserId)
        }

        // Determine test mode based on build configuration
        #if targetEnvironment(simulator)
        let testMode = true
        #elseif DEBUG
        let testMode = true
        #else
        let testMode = false
        #endif
        
        return await withCheckedContinuation { continuation in
            cloudX.initializeSDK(appKey: appKey, testMode: testMode) { success, error in
                if success {
                    print("✅ SDK Initialized (testMode: \(testMode))")
                    NotificationCenter.default.post(name: .sdkInitialized, object: nil)
                } else {
                    print("❌ SDK Init Failed: \(error?.localizedDescription ?? "Unknown error")")
                    self.showAlert(title: "SDK Init Failed", message: error?.localizedDescription ?? "Unknown error")
                }
                continuation.resume()
            }
        }
    }
    
    func updateStatusUI(state: AdState) {
        DispatchQueue.main.async { [weak self] in
            self?.statusLabel.text = state.text
            self?.statusLabel.textColor = state.color
            self?.statusIndicator.backgroundColor = state.color
        }
    }
    
    private func setupStatusUI() {
        // Setup status indicator stack
        statusStack.addArrangedSubview(statusIndicator)
        statusStack.addArrangedSubview(statusLabel)
        
        view.addSubview(statusStack)
        
        NSLayoutConstraint.activate([
            statusStack.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            statusStack.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -20),
            
            statusIndicator.widthAnchor.constraint(equalToConstant: 12),
            statusIndicator.heightAnchor.constraint(equalToConstant: 12)
        ])
    }
    
    func setupCenteredButton(title: String, action: Selector) {
        let button = UIButton(type: .system)
        button.setTitle(title, for: .normal)
        button.titleLabel?.font = UIFont.boldSystemFont(ofSize: 16)
        button.backgroundColor = .systemBlue
        button.setTitleColor(.white, for: .normal)
        button.layer.cornerRadius = 8
        button.translatesAutoresizingMaskIntoConstraints = false
        button.contentEdgeInsets = UIEdgeInsets(top: 12, left: 24, bottom: 12, right: 24)
        button.addTarget(self, action: action, for: .touchUpInside)
        view.addSubview(button)
        NSLayoutConstraint.activate([
            button.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            button.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            button.widthAnchor.constraint(equalToConstant: 200),
            button.heightAnchor.constraint(equalToConstant: 44)
        ])
    }
    
    private lazy var sdkDebuggerButton: UIButton = {
        let button = UIButton(type: .system)
        button.titleLabel?.font = UIFont.systemFont(ofSize: 14, weight: .medium)
        button.layer.cornerRadius = 6
        button.translatesAutoresizingMaskIntoConstraints = false
        button.addTarget(self, action: #selector(toggleSDKDebugger), for: .touchUpInside)
        return button
    }()
    
    func setupShowLogsButton() {
        // App Logs button (shows demo app logs)
        let showLogsButton = UIButton(type: .system)
        showLogsButton.setTitle("App Logs", for: .normal)
        showLogsButton.titleLabel?.font = UIFont.systemFont(ofSize: 14, weight: .medium)
        showLogsButton.backgroundColor = UIColor.systemOrange
        showLogsButton.setTitleColor(UIColor.white, for: .normal)
        showLogsButton.layer.cornerRadius = 6
        showLogsButton.translatesAutoresizingMaskIntoConstraints = false
        
        showLogsButton.addTarget(self, action: #selector(showLogsModal), for: .touchUpInside)
        
        view.addSubview(showLogsButton)
        view.addSubview(sdkDebuggerButton)
        
        updateSDKDebuggerButtonTitle()
        
        NSLayoutConstraint.activate([
            showLogsButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
            showLogsButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20),
            showLogsButton.widthAnchor.constraint(equalToConstant: 100),
            showLogsButton.heightAnchor.constraint(equalToConstant: 32),
            
            sdkDebuggerButton.topAnchor.constraint(equalTo: showLogsButton.bottomAnchor, constant: 8),
            sdkDebuggerButton.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20),
            sdkDebuggerButton.widthAnchor.constraint(equalToConstant: 100),
            sdkDebuggerButton.heightAnchor.constraint(equalToConstant: 32)
        ])
    }
    
    private func updateSDKDebuggerButtonTitle() {
        let isEnabled = CloudXCore.isVisualDebuggingEnabled()
        if isEnabled {
            sdkDebuggerButton.setTitle("Debugger ✓", for: .normal)
            sdkDebuggerButton.backgroundColor = .systemGreen
        } else {
            sdkDebuggerButton.setTitle("Debugger", for: .normal)
            sdkDebuggerButton.backgroundColor = .systemPurple
        }
        sdkDebuggerButton.setTitleColor(.white, for: .normal)
    }
    
    @objc private func toggleSDKDebugger() {
        let currentState = CloudXCore.isVisualDebuggingEnabled()
        CloudXCore.setVisualDebuggingEnabled(!currentState)
        updateSDKDebuggerButtonTitle()
    }
    
    @objc private func showLogsModal() {
        let logsModal = LogsModalViewController(title: "Demo App Logs")
        present(logsModal, animated: true)
    }
} 
